c++/src/capnp/serialize-async.h

```diff
@@ -139,6 +151,139 @@ class AsyncCapabilityMessageStream final: public MessageStream {
   kj::AsyncCapabilityStream& stream;
 };
 
+struct BufferedStreamOptions {
+  uint64_t defaultReadSize = 65536; // 64KiB
+  // Max amount of bytes to read in tryReadMessage if stream.tryGetLength() returns nullptr.
+
+};
+
+struct BufferedCapabilityStreamOptions : public BufferedStreamOptions {
+  uint maxFdsPerMessage;
+};
+
+namespace _ {
+
+template<typename Stream>
+extern kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessageInternal(Stream& stream,
+    uint64_t defaultReadSize, kj::ArrayPtr<kj::AutoCloseFd> passedInFdSpace,
+    ReaderOptions readerOptions, kj::ArrayPtr<word> scratchSpace);
+
+} // namespace _
+
+class BufferedAsyncIoMessageStream final : public MessageStream {
+public:
+  explicit BufferedAsyncIoMessageStream(kj::AsyncIoStream& stream,
+      BufferedStreamOptions options = BufferedStreamOptions{})
+      : stream(stream), options(options) {}
+
+  // Implements MessageStream
+  kj::Promise<kj::Maybe<MessageReaderAndFds>> tryReadMessage(
+      kj::ArrayPtr<kj::AutoCloseFd> fdSpace,
+      ReaderOptions options = ReaderOptions(), kj::ArrayPtr<word> scratchSpace = nullptr) override;
+  kj::Promise<void> writeMessage(
+      kj::ArrayPtr<const int> fds,
+      kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) override;
+  kj::Promise<void> writeMessages(
+      kj::ArrayPtr<kj::ArrayPtr<const kj::ArrayPtr<const word>>> messages) override;
+  kj::Maybe<int> getSendBufferSize() override;
+  kj::Promise<void> end() override;
+
+private:
+  kj::AsyncIoStream& stream;
+  BufferedStreamOptions options;
+
+  struct IncompleteRead {
+    kj::Array<kj::byte> bytes;
+    size_t expectedSize;
+  };
+
+  std::deque<kj::Own<MessageReader>> bufferedReaders;
+  kj::Maybe<IncompleteRead> incompleteRead;
+
+  // Implementation details used by tryReadMessageInternal
+  kj::Maybe<MessageReaderAndFds> tryGetBufferedMessage(kj::ArrayPtr<kj::AutoCloseFd> fdSpace);
+  kj::Maybe<IncompleteRead> tryGetIncompleteRead();
+  void bufferCompleteMessage(MessageReaderAndOwnedFds message);
+  void bufferIncompleteMessage(kj::ArrayPtr<kj::byte> leftover, kj::Array<kj::AutoCloseFd> fds, size_t size);
+
+  struct ReadResult {
+    kj::Array<kj::byte> buffer;
+    size_t byteCount;
+
+    kj::Array<kj::AutoCloseFd> getFds(bool isLast) { return nullptr; }
+  };
+  kj::Promise<ReadResult> tryRead(kj::Array<byte> buffer, void* start, size_t minBytes,
+      size_t maxBytes);
+
+  friend kj::Promise<kj::Maybe<MessageReaderAndFds>>
```

Comment:
something must be different in c++17/21 that's making this compile for me -- I'm trying some things, I'd really like to keep the implementation of tryReadMessageInternal out of the header